### ISSUE TITLE
Multi-line error highlights make it hard to work

### ISSUE THREAD

**Issue by DamienCassou** [flycheck/flycheck#1730]
## Checklist
- [x] I have checked existing issues for potential duplicates before creating this one.
- [x] I have read the [Troubleshooting guide][].

## Bug description

It's not a bug, rather an annoying feature.

![2020-04-10-104912](https://user-images.githubusercontent.com/217543/78977853-2302b400-7b19-11ea-8b5d-85cdd086bb01.png)

Flycheck correctly marks the code above as being problematic. Eslint complains similarly:

```
$ eslint country-tests.js 

country-tests.js
  3:1  error  Unexpected fdescribe  jasmine/no-focused-tests

✖ 1 problem (1 error, 0 warnings)
```

The [jasmine/no-focused-tests](https://github.com/tlvince/eslint-plugin-jasmine/blob/HEAD/docs/rules/no-focused-tests.md) rule discourages the use of `fdescribe()` unless during development. Because everything is underlined in flycheck now, using `fdescribe()` becomes nearly impossible even during development.

## Expected behavior

I think the previous behavior of only marking the first letter of `fdescribe` was more usable.

## System configuration
```
  javascript-eslint
    - may enable:  yes
    - executable:  Found at /usr/local/bin/eslint_d
    - config file: found


--------------------

Flycheck version: 32snapshot
Emacs version:    27.0.90
System:           x86_64-pc-linux-gnu
Window system:    x
```

Emacs configuration:
- [x] Plain Emacs / Custom configuration
- [ ] Spacemacs
- [ ] Doom Emacs
- [ ] Other shared configuration

**Comment by cpitclaudel** [comment]
Sorry for the trouble.  I think this *is* a bug, but on the `jasmine/no-focused-tests` side. We're just respecting the range that it returns, which is way to big (as you suggest, highlighting just one word would be perfect).  We can work around this in Flycheck, but could you also report it to them?

**Comment by cpitclaudel** [comment]
I wonder if we should put a cap on the number of highlighted lines

**Comment by cpitclaudel** [comment]
Maybe we ought to have a variable at least that says whether to use spans.

**Comment by DamienCassou** [comment]
> could you also report it to them?

done: https://github.com/tlvince/eslint-plugin-jasmine/issues/230

> I wonder if we should put a cap on the number of highlighted lines

do you know of any case where highlighting more than 1 line is meaningful?

**Comment by cpitclaudel** [comment]
> done: tlvince/eslint-plugin-jasmine#230

Brilliant :clap: 

> do you know of any case where highlighting more than 1 line is meaningful?

Yes, unfortunately; here's an OCaml example where we currently do terrible, and supporting (multi-line) ranges will help (I haven't pushed multi-line support to the flycheck-ocaml checker yet):

```ocaml
let f (x: int) =
  match x with
  | 0 -> "a"
  | _ -> (+)
           1 3
```

Right now we highlight the first character of `(+)` and say `This expression has type int but an expression was expected of type string`, which is quite confusing.

Here's a Coq example showing a similar issue:

```coq
Definition a :=
  True \/
  False
    :: [].
```

This parses as `Definition a := True \/ (False :: []).`

Which is ill-typed; the error spans the whole `False :: []` term; if we highlighted just `False` (the first line), we'd be suggesting the wrong thing.

But indeed, giant error spans are a problem.  In fact, I turned multi-line support off entirely in phpmd because it only highlighted huge spans…

Maybe we should find less intrusive ways to report errors, so that it wouldn't be such a problem to develop even with large error spans highlighted.  But you seem to have customized flycheck to highlight the text in red in addition to underlining it, which goes against that idea.  WDYT?

**Comment by vxcamiloxv** [comment]
Same issue with eslint in a normal JS file with js2-mode and web-mode

**Comment by vxcamiloxv** [comment]
After some tests looks like the issue was introduces in https://github.com/flycheck/flycheck/commit/936dbf52a223244eee8fe2b26efed4a3b6379331#diff-374cdd9563b826c712b1dc6a373a0bc1, before that commit all works good to me

**Comment by cpitclaudel** [comment]
> After some tests looks like the issue was introduces in 936dbf5#diff-374cdd9563b826c712b1dc6a373a0bc1, before that commit all works good to me

Can you double-check? I don't think that commit is related.
Can you show the source file?

**Comment by vxcamiloxv** [comment]
Yes I confirm the issue with https://github.com/flycheck/flycheck/blob/936dbf52a223244eee8fe2b26efed4a3b6379331/flycheck.el, 

Works fine with the previous commit code: https://github.com/flycheck/flycheck/blob/7be341c24a235281ef4e5b87787bff35a23007c6/flycheck.el

You can test it with a small code:

```javascript
// date-test.js
/**
 * Helper for creating a formatted date string
 * @param {date} d Date object
 * @param {string} lang Language (only supports English 'en' and Spanish 'es')
 * @param {string} timeZone Time zone to use, defaults to America/New_York. Will be ignored in
 *                          unsupported browsers.
 * @returns {string} Spanish formatted date (ie: '18 Jul | 06:29 EDT')
 */
export function formatDate(d, lang = 'es', timeZone = 'America/New_York') {
  const time = timeZone;
  const order = lang;

  return `${order} – ${time}`;
}
``` 
and this eslint plugin: https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/prefer-default-export.md
```eslint 
extends:
  - eslint:recommended
plugins:
  - import
rules:
  import/prefer-default-export: 2
```

**Comment by DamienCassou** [comment]
> Maybe we should find less intrusive ways to report errors, so that it wouldn't be such a problem to develop even with large error spans highlighted. But you seem to have customized flycheck to highlight the text in red in addition to underlining it, which goes against that idea. WDYT?

I'm using [modus-operandi theme](https://gitlab.com/protesilaos/modus-themes) from @protesilaos. I guess the themes could be improved to highlight that there is a problem without making it hard to fix it.

**Comment by protesilaos** [comment]
Thanks for noting this!  I will review them right now and report back.

**Comment by cpitclaudel** [comment]
@vxcamiloxv thanks.  As in Damien's case, this is partly an upstream bug: eslint reports a giant error range, so we obediently highlight a giant portion of the buffer.  Can you report this issue to the maintainer of the eslint plugin?

…but I agree that, in the meantime, we should find a fix, since this isn't very convenient.

I can think of a few options:

* Switching to a different highlighting style, such as inserting markers instead of highlighting, something `→like this←` or `⚠️[like this]`, or even using the indicators we currently put in the margin, `»like this«`

* Putting a bound on the number of highlighted characters, so that we highlight at most, say, 30 characters; but then we'd need to chose how to clarify that the error goes on.

* Adding a setting to disable region highlighting fully.

etc.  I'd love to hear ideas and opinions.

**Comment by DamienCassou** [comment]
or a combination: underline if the error is on a single line and have markers in case the error spans several lines. Markers have the advantage of allowing multiple errors in the same region such as a small error inside an error spanning several lines.

**Comment by vxcamiloxv** [comment]
@cpitclaudel I'll report the bug also to the upstream eslint plugin

**Comment by cpitclaudel** [comment]
Hi all,

I've made a bit of progress on this.  Here are some mock-ups:

* Delimiters around the error span:
    ![delimiters](https://user-images.githubusercontent.com/8181630/79895380-5507f600-83f6-11ea-8352-1e6bfc96dfce.png)

* Fringe indicators 
    ![fringes](https://user-images.githubusercontent.com/8181630/79895386-55a08c80-83f6-11ea-8131-26f46d35e408.png)

The idea would be to show regular underlines + fringe indicators up to a configurable number of lines, and then switch to paired delimiters + fringe indicators for longer errors.

Now I need to pick a default delimiter and fringe style.  Do you have preferences among the above?

**Comment by NicolasPetton** [comment]
@cpitclaudel In terms of UX, don't you think it could be distracting/annoying to have delimiters shown in the code, as they will move text horizontally on the line?

**Comment by cpitclaudel** [comment]
@NicolasPetton Definitely, yes.  Especially with transient errors that appear and disappear as you type.  From my experience programming with it for a few hours, the most annoying part is the start marker, because that moves the place where the point is as well (whereas the end of the error tends to be after the point).  On the other hand, it's quite nice to see nested errors.

Can you think of something else that would be more discrete? (I'm fairly happy with the continuation lines in the fringe, but I don't think these are enough on their own).

**Comment by cpitclaudel** [comment]
Also, thanks for having a look!

**Comment by DamienCassou** [comment]
Thank you very much for your work. Isn't it the responsibility of the theme to decide how to show something?

**Comment by NicolasPetton** [comment]
> Can you think of something else that would be more discrete? (I'm fairly happy with the continuation lines in the fringe, but I don't think these are enough on their own).

What about simply underlining the first line? With the indications in the fringe, wouldn't it be enough?

**Comment by cpitclaudel** [comment]
> Thank you very much for your work. Isn't it the responsibility of the theme to decide how to show something?

I don't think Emacs themes have this level of control, and in any case we need to come up with good defaults.  But in any case, this stuff would be configurable.

**Comment by cpitclaudel** [comment]
> What about simply underlining the first line? With the indications in the fringe, wouldn't it be enough?

Interesting. That's a heavier change because we'd need two overlays instead of one, but we could do that.  That being said, there's still the problem of indicating the end of the error.

**Comment by protesilaos** [comment]
> Thank you very much for your work. Isn't it the responsibility of the theme to decide how to show something?

A theme is supposed to control the "faces" themselves (`defface`), but not how they are applied to interface/code constructs.  So in this case the theme is responsible for the colour and style of the underline (straight line or wave), the colour of the affected text, and the styles of the fringe indicators.

**Comment by fmdkdd** [comment]
What about showing the full error span only for the currently highlighted error (the one that triggered `-display-error`)?  The full span may be useful to understand the error, but it doesn't need to be displayed the rest of the time (when navigating or editing).

**Comment by cpitclaudel** [comment]
When I'm typing text with a fast-enough checker parts of the current line are often highlighted, so I worry that this too will cause issues.

**Comment by DamienCassou** [comment]
>  That being said, there's still the problem of indicating the end of the error.

Do we really have to?

**Comment by cpitclaudel** [comment]
> Do we really have to?

Good question.  I don't have a very strong opinion.  It can help to see where the error end, but if it messes up the layout it might not make sense.

**Comment by cpitclaudel** [comment]
OK, I've cleaned up my branch and added documentation.  Please see this PR: https://github.com/flycheck/flycheck/pull/1743.  It would be great if you could have a look and ideally give it a try (comments on the design are also very welcome).  I've defaulted to hiding underlines when the error spans more than 4 lines.  The threshold is customizable and delimiters can be added if desired.
